sensors/movementsensor.go

```diff
@@ -0,0 +1,324 @@
+type TimedIMUReadingResponse struct {
+	AngularVelocity    spatialmath.AngularVelocity
+	LinearAcceleration r3.Vector
+	ReadingTime        time.Time
+}
+
+// TimedOdometerReadingResponse represents an odometer sensor reading with a time.
+type TimedOdometerReadingResponse struct {
+	Position    *geo.Point
+	Orientation spatialmath.Orientation
+	ReadingTime time.Time
+}
+
+// MovementSensor represents a movement sensor.
+type MovementSensor struct {
+	name              string
+	dataFrequencyHz   int
+	imuSupported      bool
+	odometerSupported bool
+	replay            bool
+	sensor            movementsensor.MovementSensor
+}
+
+// Name returns the name of the movement sensor.
+func (ms *MovementSensor) Name() string {
+	return ms.name
+}
+
+// DataFrequencyHz returns the data rate in ms of the movement sensor.
+func (ms *MovementSensor) DataFrequencyHz() int {
+	return ms.dataFrequencyHz
+}
+
+// TimedMovementSensorReading returns data from the movement sensor and the time the reading is from & whether
+// it was a replay sensor or not.
+func (ms *MovementSensor) TimedMovementSensorReading(ctx context.Context) (TimedMovementSensorReadingResponse, error) {
+	var (
+		readingTimeAngularVel, readingTimeLinearAcc time.Time
+		readingTimePosition, readingTimeOrientation time.Time
+		angVel                                      spatialmath.AngularVelocity
+		linAcc                                      r3.Vector
+		position                                    *geo.Point
+		orientation                                 spatialmath.Orientation
+		timedIMUReadingResponse                     *TimedIMUReadingResponse
+		timedOdometerReadingResponse                *TimedOdometerReadingResponse
+		err                                         error
+	)
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, timedMovementSensorReadingTimeout)
+	defer cancel()
+	if ms.imuSupported {
+	imuLoop:
+		for {
+			select {
+			case <-timeoutCtx.Done():
+				return TimedMovementSensorReadingResponse{}, timeoutCtx.Err()
+			default:
+				if timedIMUReadingResponse, err = ms.timedIMUReading(timeoutCtx, &angVel, &linAcc,
+					&readingTimeAngularVel, &readingTimeLinearAcc); err != nil && !errors.Is(err, ErrNoValidReadingObtained) {
+					return TimedMovementSensorReadingResponse{}, err
+				}
+				if timedIMUReadingResponse != nil {
+					break imuLoop
+				}
+			}
+		}
+	}
+	if ms.odometerSupported {
+	odometerLoop:
+		for {
+			select {
+			case <-timeoutCtx.Done():
+				return TimedMovementSensorReadingResponse{}, timeoutCtx.Err()
+			default:
+				if timedOdometerReadingResponse, err = ms.timedOdometerReading(timeoutCtx, position, &orientation,
+					&readingTimePosition, &readingTimeOrientation); err != nil && !errors.Is(err, ErrNoValidReadingObtained) {
+					return TimedMovementSensorReadingResponse{}, err
+				}
+				if timedOdometerReadingResponse != nil {
+					break odometerLoop
+				}
+			}
+		}
+	}
+	return TimedMovementSensorReadingResponse{
```

Comment:
Should we handle the case where imu and odom were requested, odom fails but imu succeeds (and vica versa)?

Reply:
This would be resolved if we use Readings in the near future (ticket: https://viam.atlassian.net/browse/RSDK-5736). In that case, getting the reading from the movement sensor would fail as well when one of the endpoints fails. Which is why I think that using the approach I implemented is fine.

Reply:
Agreed